File: 0099-recover-binary-search-tree/0099-recover-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void inorder(TreeNode* root, TreeNode** prev, TreeNode** first, TreeNode** middle, TreeNode** last){
        if(root==NULL){
            return;
        }
        inorder(root->left, prev, first, middle, last);
        if(*prev!=NULL && (*prev)->val > root->val){
            if(*first==NULL){
                *first = *prev;
                *middle = root;
            }
            else{
                *last = root;
            }
        }
        *prev = root;
        inorder(root->right, prev, first, middle, last);
    }
    void recoverTree(TreeNode* root) {
        if(root==NULL){
            return;
        }
        TreeNode* prev = new TreeNode(INT_MIN);
        TreeNode* first = NULL;
        TreeNode* middle = NULL;
        TreeNode* last = NULL;
        inorder(root, &prev, &first, &middle, &last);
        if(first!=NULL && last!=NULL){
            int temp = first->val;
            first->val = last->val;
            last->val = temp;
        }
        else if(first!=NULL && middle!=NULL){
            int temp = first->val;
            first->val = middle->val;
            middle->val = temp;
        }
    }
};

File: 0099-recover-binary-search-tree/0099-recover-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void inorder(TreeNode root, TreeNode[] prev, TreeNode[] first, TreeNode[] middle, TreeNode[] last) {
        if (root == null) {
            return;
        }
        inorder(root.left, prev, first, middle, last);
        if (prev[0] != null && prev[0].val > root.val) {
            if (first[0] == null) {
                first[0] = prev[0];
                middle[0] = root;
            } else {
                last[0] = root;
            }
        }
        prev[0] = root;
        inorder(root.right, prev, first, middle, last);
    }

    public void recoverTree(TreeNode root) {
        if (root == null) {
            return;
        }
        TreeNode[] prev = {null};
        TreeNode[] first = {null};
        TreeNode[] middle = {null};
        TreeNode[] last = {null};
        inorder(root, prev, first, middle, last);
        if (first[0] != null && last[0] != null) {
            int temp = first[0].val;
            first[0].val = last[0].val;
            last[0].val = temp;
        } else if (first[0] != null && middle[0] != null) {
            int temp = first[0].val;
            first[0].val = middle[0].val;
            middle[0].val = temp;
        }
    }
}